### DISCUSSION TITLE
Is it a common practice to implement complex interactions using Global Callbacks?

### DISCUSSION THREAD

**Discussion by wspl** [slint-ui/slint#2685] [Q&A]
I'm creating a ListView that supports multiple selection. It requires manipulating arrays (such as `find`, `insert`, `filter`) in various scenarios. To implement these manipulation functions, I use Global Callbacks and pass the arrays from the slint file to native code.

I find this approach interesting but somewhat tricky, as my functions can only be used in this specific scenario (no generic type, and I need to pass flags to the callback). It makes it difficult to extend them to other situations. Additionally, I can't use the slint viewer.

So, I'm curious if my practice is common, or if there's a better way to achieve this?

**Comment by ogoffart** [comment]
Yes, this is common practice.
The other way would be to forward the callback with aliases up to the component root, but this is more boilerplate.

We are looking to implement better way to handle it.
https://github.com/slint-ui/slint/issues/2031  or https://github.com/slint-ui/slint/issues/251
But currently the best we have is using global

**Reply by wspl** [reply]
Perhaps we can introduce a lightweight scripting capability in the slint file, such as embedding a quickjs, so that we can operate on arrays and handle some complex string logic.

I think this might increase the complexity of the slint file, but it should reduce the complexity of the overall program architecture.

**Reply by ogoffart** [reply]
The idea is to split the logic and UI between the native code and the .slint files.  The .slint fiels shouldn't contain much logic because it only represents the UI.

**Reply by wspl** [reply]
I agree with and appreciate the principle that the slint file should only contain code related to the UI.

Suppose I'm developing a UI library that includes a List component, which is based on some Rust logic. To distribute this UI library, besides distributing the slint file, I would also need to distribute a Rust crate to ensure users can use it properly. For C++ users, I might need to write a C++ version to implement this logic.

In this case, does slint ui have any plans to address this issue?
